Fusion Extension implementation - inter-species

FilipO28555 opened the pull request:
This extension is based on papers:
[Wu et al](https://doi.org/10.1063/5.0051178)
[PIC code Chicago](https://www.sciencedirect.com/science/article/pii/S0021999119302037)
[Cannoni](https://arxiv.org/abs/1605.00569v2)
Cross section parametrization:
[Bosh, Hale](https://iopscience.iop.org/article/10.1088/0029-5515/32/4/I07)

There is now documentation.

Code flow:
 ```
                                                                ┌ IntraCollision.hpp (not yet tested/included - will be in the next PR)
 simulation.hpp -> Fusion.x.cpp -> Collider.hpp -> WithPeer.hpp ┤
                                                                └ InterCollision.hpp
```
Struct `InterCollision{}` is a combination of algorithm from Collision Extension and Creation Kernel. 
```
                   uses: (particles/fusion/detail) 
InterCollision.hpp --------------------------------> FusionFunctor.hpp -> FusionAlgorithm.hpp
```
`FusionAlgorithm.hpp` decides if fusion happens and chooses momenta of new particles.

The setup that works and tests this extension is at: [Fusion_Setups/DT-nHe_hal8999](https://github.com/FilipO28555/Fusion_Setups/tree/DT-nHe_hal8999)

Some preliminary results:
![Results/predicted_vs_measured_by_ppc.png](https://raw.githubusercontent.com/FilipO28555/Fusion_Setups/DT-nHe_hal8999/Results/predicted_vs_measured_by_ppc.png)
![Results/predicted_vs_measured_by_ppc.png](https://raw.githubusercontent.com/FilipO28555/Fusion_Setups/DT-nHe_hal8999/Results/ppc_error_analysis.png)

Added CI test! And documentation references its fusion.param file.

In the next PR:
 - Add intra-species fusion
 - More cross-section interpolators

FilipO28555 committed:
Add fusion implementation DOI:/10.1063/5.0051178
Features:
 - binary fusion reactions between two different species
 - choosing product weights depending on particles masses and charge.
 - documentation

BrianMarre commented on include/picongpu/param/fusion.param:
```diff
@@ -0,0 +1,106 @@
+
+            using namespace picongpu::plugins;
+
+            // not ideal, better would be to combine filters with an AND but it is ok
```
what does this comment refer to?

FilipO28555 replied:
it was a comment from collision param. I touched up the entire file now

BrianMarre commented on include/picongpu/param/fusion.param:
```diff
@@ -0,0 +1,106 @@
+                        static constexpr float_X B1 = 6.38e1_X;
+                        static constexpr float_X B2 = -9.95e-1_X;
+                        static constexpr float_X B3 = 6.981e-5_X;
+                        static constexpr float_X B4 = 1.728e-4_X;
```
you have a python script for determining the reaction cross section parameters if I remember correctly.
Please include it in this PR.

pordyna commented:
Right now this reuses a lot of collision code simply by copying files and also somewhat modifying them. We should unify this, for example, by generalizing the collision kernels. I'm not sure if this should happen before this is merged. @psychocoderHPC what do you think?

psychocoderHPC commented:
> Right now this reuses a lot of collision code simply by copying files and also somewhat modifying them. We should unify this, for example, by generalizing the collision kernels. I'm not sure if this should happen before this is merged. @psychocoderHPC what do you think?

we should keep the code first seperate, after merging we have a baseline to validate if we break something. Than we should check what we can unify.
This way requires that we not stop after after it is merged.
Before next week I am most likely not possible to look into this pr.

BrianMarre commented:
@FilipO28555 the CI is still complaining that some source code files are marked as executable, for example,

- include/picongpu/particles/fusion/detail/ListEntry.hpp
- include/picongpu/particles/fusion/WithPeer.hpp

only binaries and scripts should be marked as executable, please follow the directions in the CI-job output on how to fix this.

FilipO28555 committed:
Improved documentation

PrometheusPi commented on include/picongpu/param/fusion.param:
```diff
@@ -0,0 +1,192 @@
+             * parameterization from experimental data.
+             * Source: https://hdl.handle.net/11858/00-001M-0000-0027-6535-1 (page 29)
+             */
+            // struct DT
```
Please remove uncommented /unused code

ikbuibui replied:
One option if you want to keep this code is to add it as an example and activate it. That way we get better CI testing as well.

FilipO28555 replied:
OK, Now the fusion.param is a bare minimum implementation - I moved the examples to documentation

ikbuibui replied:
I see you have really moved all the example code into the documentation. 
We avoid doing this because it is easy to overlook updating the docs when something in your implementation changes, and we end up with code in our docs which doesn't compile. 
The way we deal with this is to put the param file into an example/test set up under `share/picongpu/examples` or `share/picongpu/tests` and then include this code into the doxygen. This way the code is checked for compilation and if something in the code changes, you don't have to separately updated the docs.
You can look at [this](https://github.com/ComputationalRadiationPhysics/picongpu/blob/e2c179b91072070a90cad1ddee87a15901ca7116/docs/source/usage/plugins/binningPlugin.rst?plain=1#L124)  to see how to include code into doxygen.

FilipO28555 replied:
done

PrometheusPi reviewed (changes requested):
@FilipO28555 This is not yet a full review, but it would be great if you remove unused /comment-out code so that the number of added files gets less scary (right now it's 3.5k lines)

ikbuibui commented:
In general, we avoid merge commits
See [here](https://picongpu.readthedocs.io/en/latest/dev/CONTRIBUTING.html#keep-track-of-updates) and [here](https://picongpu.readthedocs.io/en/latest/dev/docs/COMMIT.html#rebasing-onto-a-change-of-convention-formatting)

ikbuibui commented:
Also I'm not sure how/if possible it is anymore but it would have been useful to have a commit at the state where you simply copied code from the collision pipeline and made it compile, and then added the fusion changes on top with separate commits.

FilipO28555 committed:
Clean up fusion.param and improve documentation

- Remove all verbose comments and examples from fusion.param
- Move complete D-T fusion example to documentation

BrianMarre commented:
@psychocoderHPC any progress on reviewing this PR?

psychocoderHPC commented on include/picongpu/param/fusion.param:
```diff
@@ -0,0 +1,54 @@
+
+#pragma once
+
+#include "picongpu/param/particleFilters.param"
```
Please do not include a param in a param file.

Add  a file `param.hpp` into `particles/fusion` and handle the dependencies there see https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/include/picongpu/particles/collision/param.hpp

psychocoderHPC commented on include/picongpu/param/fusion.param:
```diff
@@ -0,0 +1,54 @@
+        {
+            namespace precision
+            {
+                using float_COLL = float_64;
```
copy past issue: please remove `float_COLL` if you not use it or rename ist e.g. to `float_fustion` or `float_Fusion`, we do not have an exact naming schema for this kind.

psychocoderHPC commented on include/picongpu/param/fusion.param:
```diff
@@ -0,0 +1,54 @@
+            /** Define your fusion reactions here.
+             * See documentation: https://picongpu.readthedocs.io/en/latest/usage/workflows/fusionReactions.html
+             */
+
```
please remove the empty line

psychocoderHPC commented on include/picongpu/param/fusion.param:
```diff
@@ -0,0 +1,54 @@
+
+            constexpr uint32_t cellListChunkSize = TYPICAL_PARTICLES_PER_CELL;
+            constexpr float_X productMinWeighting = 16.1;
+            constexpr uint32_t maxFmult = 1e6;
```
please add a doc string and unit, additionally the type is an integral type but you assigned an double value.

psychocoderHPC commented on include/picongpu/param/fusion.param:
```diff
@@ -0,0 +1,54 @@
+            using FusionPipeline = pmacc::mp_list<>;
+
+            constexpr uint32_t cellListChunkSize = TYPICAL_PARTICLES_PER_CELL;
+            constexpr float_X productMinWeighting = 16.1;
```
please add a doc string

psychocoderHPC commented on include/picongpu/particles/fusion/fieldSlots.hpp:
```diff
@@ -0,0 +1,33 @@
+        namespace fusion
+        {
+            //! FieldTmp slot which is used to store the calculated Debye length
+            constexpr uint32_t screeningLengthSlot = 0u;
```
If I not missed something this variable is not used, therefore the file can be removed.

psychocoderHPC commented on include/picongpu/particles/fusion/detail/Creation.hpp:
```diff
@@ -0,0 +1,455 @@
+            T truncated = static_cast<long long>(x);
+            T fraction = x - truncated;
+
+            if(fraction >= 0.5)
```
should be `T{0.5}` to avoid silent precision switch. The same for all other hard coded values, also in other files if exits.
`0.5` only is always double in c++.

psychocoderHPC commented on include/picongpu/particles/fusion/detail/Creation.hpp:
```diff
@@ -0,0 +1,455 @@
+        template<typename T>
+        constexpr T round_cx(T x)
+        {
+            T truncated = static_cast<long long>(x);
```
should be `T truncated = math::floor(x);`

psychocoderHPC replied:
or use `math::trunc(x)` see https://github.com/ComputationalRadiationPhysics/picongpu/blob/dev/include/pmacc/math/functions/Round.hpp

FilipO28555 replied:
@psychocoderHPC Can't use this because it's not constexpr.

steindev replied:
ping @psychocoderHPC

psychocoderHPC commented on include/picongpu/particles/fusion/detail/Creation.hpp:
```diff
@@ -0,0 +1,455 @@
+            float_COLL c4)
+        {
+            // clang-format off
+            return (W1 >= 0.0_X && W1 <= c3 &&
```
`0.0_X` is not nessesary `float_COLL` you should always write `float_COLL{0.0}` to keep the precision.

psychocoderHPC commented on include/picongpu/particles/fusion/detail/Creation.hpp:
```diff
@@ -0,0 +1,455 @@
+            float_COLL m4)
+        {
+            constexpr float_COLL tolerance
+                = static_cast<precision::float_COLL>(10) * std::numeric_limits<float_COLL>::epsilon();
```
`static_cast<precision::float_COLL>(10)` can be `precision::float_COLL{10}`

psychocoderHPC replied:
or simply `10.0_COLL`

psychocoderHPC commented on include/picongpu/particles/fusion/detail/Creation.hpp:
```diff
@@ -0,0 +1,455 @@
+            {
+                float_COLL const c3 = (Q * m4 - q4 * M) / det;
+                float_COLL const c4 = (q3 * M - Q * m3) / det;
+                bool const ok = (c3 >= 0.0_X && c4 >= 0.0_X);
```
same here `0.0` should be the corerct precision

psychocoderHPC commented on include/picongpu/particles/fusion/detail/Creation.hpp:
```diff
@@ -0,0 +1,455 @@
+                IdGenerator& idGen,
+                T_ParAccessor0 const& r1,
+                T_ParAccessor1 const& r2,
+                float_X& productWeighting,
```
`float_X&` should be `float_X` because it is used as input only. `float_X const&` would be fine too.

psychocoderHPC commented on include/picongpu/particles/fusion/detail/Creation.hpp:
```diff
@@ -0,0 +1,455 @@
+                T_ParAccessor0 const& r1,
+                T_ParAccessor1 const& r2,
+                float_X& productWeighting,
+                float3_X& mom1,
```
This and next line should be `float3_X const&`

psychocoderHPC reviewed (changes requested):
Please review your own changes a remove hard coded numbers e.g. `1.0` by the correct precision of the value e.g. `1.0_COLL` to avoid silent implicit casts.

FilipO28555 committed:
Got rid of unnecessary file. Unified variable type declaration.

BrianMarre commented:
@psychocoderHPC a few CI jobs failed with due to script errors, please restart them

PrometheusPi commented:
CI job restarted

PrometheusPi commented:
I triggered the CI again, we have some connectivity issues: 
> fatal: unable to access 'https://github.com/ComputationalRadiationPhysics/isaac.git/': Could not resolve host: github.com

FilipO28555 committed:
Removed parallel max finding algorithm because it contained a bug. Changed logic of Fusion multiplier parameter to be better (updated documentation as well). Added some debug printouts.

FilipO28555 committed:
Added CI test and updated documentation to referance fusion.param file in it

PrometheusPi commented:
What is the status of this PR?

BrianMarre commented:
Needs to be reviewd by @psychocoderHPC or @ikbuibui together with @FilipO28555

PrometheusPi commented:
@psychocoderHPC @ikbuibui @steindev what's the status of your review with @FilipO28555 ?

steindev commented:
> @psychocoderHPC @ikbuibui @steindev what's the status of your review with @FilipO28555 ?

It was brought to my attention again today. Will review this week.

steindev commented on docs/source/usage/workflows/fusionReactions.rst:
```diff
@@ -0,0 +1,147 @@
+.. _fusionReactions:
+
+Fusion Workflow
```
Better call it 'Nuclear Fusion Reactions'

steindev commented on docs/source/usage/workflows/fusionReactions.rst:
```diff
@@ -0,0 +1,147 @@
+
+Simulation Parameters
+---------------------
+
```
Please add a comment explaining in which file the following code snippet needs to be put.

steindev commented on include/picongpu/particles/fusion/detail/arrayHelpers.hpp:
If this file already exists somewhere else, you should not duplicate it but use the original.

steindev commented on include/picongpu/particles/fusion/detail/cellDensity.hpp:
If this file already exists somewhere else, you should not duplicate it but use the original.

steindev commented on include/picongpu/particles/fusion/detail/Creation.hpp:
```diff
@@ -0,0 +1,452 @@
+            bool isValid;
+        };
+
+        // New: stoichiometric caps for products 1 and 2
```
How long is this supposed to stay 'new'? ;-)

steindev commented on include/picongpu/particles/fusion/detail/Creation.hpp:
```diff
@@ -0,0 +1,452 @@
+        };
+
+        // New: compute stoichiometric caps (c3, c4) from global charge and mass conservation
+        constexpr StoichiometryCaps computeStoichiometryCaps(
```
How long is this supposed to stay 'new'?

steindev commented on include/picongpu/particles/fusion/detail/Creation.hpp:
```diff
@@ -0,0 +1,452 @@
+            return StoichiometryCaps{0.0_COLL, 0.0_COLL, false};
+        }
+
+        // New: local mass+charge solver under caps. Solve site-1 2x2 and derive site-2 via caps.
```
What is site-1 and site-2?

steindev commented on include/picongpu/particles/fusion/relativistic/FusionFunctor.def:
```diff
@@ -0,0 +1,43 @@
+                template<typename T_Param>
+                struct FusionFunctorImpl;
+
+                /** Host side binary collision functor for using a constant coulomb logarithm value
```
I think, you should at least update the documentation of this `FusionFunctor` and not copy-paste this from collisions as well...

steindev commented on include/picongpu/particles/fusion/relativistic/FusionFunctor.hpp:
```diff
@@ -0,0 +1,103 @@
+    {
+
+        // clang-format off
+        //! Compile-time Cross-section functor - Pade Polynomial
```
Any reference for this computation available?

steindev commented on include/picongpu/particles/fusion/relativistic/FusionFunctor.hpp:
```diff
@@ -0,0 +1,103 @@
+
+        HINLINE FusionFunctorImpl(uint32_t currentStep) {};
+
+        using AccFunctorImpl = acc::FusionAlg<acc::CalcCrossSection<T_Param>>;
```
This looks like a good place to shortly document what the following entities do / are used for.

steindev commented on include/picongpu/particles/fusion/IBinary.def:
```diff
@@ -0,0 +1,48 @@
+             *          and support at least two particles
+             */
+            template<typename T_BinaryFunctor>
+            using IBinary = pmacc::functor::Interface<T_BinaryFunctor, 2u, void>;
```
You are not using this, aren't you?

steindev commented on include/picongpu/particles/fusion/WithPeer.hpp:
```diff
@@ -0,0 +1,158 @@
+        {
+            namespace detail
+            {
+                // in-between species fusion
```
What is this doing? I cannot make any sense out of the name and the comment.

steindev commented on include/picongpu/particles/fusion/WithPeer.hpp:
```diff
@@ -0,0 +1,158 @@
+                    colliderId,
+                    pairId>
+                {
+                    // implementation in later pull request
```
Please write this as doxygen with `@todo` (or so).

steindev commented on include/picongpu/unitless/simulation.unitless:
```diff
@@ -500,6 +518,7 @@ namespace picongpu
         static constexpr SiUnits si{};
         static constexpr Units unit{};
         static constexpr AtomicUnits atomicUnit{};
+        static constexpr FusionUnits fusion{};
```
I would vote for calling this `fusionUnit{}`.

And in general, isn't the electron mass in AMU already available in atomicPhysics and can be used from there?

steindev commented on share/picongpu/tests/Fusion/bin/ci.sh:
```diff
@@ -0,0 +1,126 @@
+  echo "Generated number of He4 particles is compared to analytical prediction."
+  echo ""
+  echo "Usage:"
+  echo "    (1) Change current working directory to direcectory where the include directory of the setup is located"
```
typo 'directory'

steindev commented on share/picongpu/tests/Fusion/bin/ci.sh:
```diff
@@ -0,0 +1,126 @@
+  TBG_n_Bin="--n_energyHistogram.period 10 --n_energyHistogram.filter all --n_energyHistogram.binCount 3 --n_energyHistogram.minEnergy 0 --n_energyHistogram.maxEnergy 2"
+  TBG_He4_Bin="--He4_energyHistogram.period 10 --He4_energyHistogram.filter all --He4_energyHistogram.binCount 3 --He4_energyHistogram.minEnergy 0 --He4_energyHistogram.maxEnergy 2"
+
+  # run the sim: 1 node, 24x24x24 cells, 250 steps
```
500 steps

steindev commented on share/picongpu/tests/Fusion/include/picongpu/param/density.param:
Please remove all the profiles you are not using. That is, all except Homogenous.

steindev commented on share/picongpu/tests/Fusion/include/picongpu/param/fusion.param:
```diff
@@ -0,0 +1,122 @@
+#include <pmacc/meta/conversion/MakeSeq.hpp>
+
+
+#ifndef PARAM_DYNAMIC_LOG
```
You don't use that, don't you?

steindev commented on share/picongpu/tests/Fusion/include/picongpu/param/fusion.param:
```diff
@@ -0,0 +1,122 @@
+                {
+                    //! Gamow constant: BG = pi*alpha*Z1*Z2*sqrt(2*mu*c^2) in keV^(1/2)
+                    //! where alpha is the fine structure constant, Z1,Z2 are charges, mu is reduced mass
+                    static constexpr float_X BG = 34.3827_X;
```
Shouldn't this be `_COLL` here and in the following definitions?

steindev commented on share/picongpu/tests/Fusion/include/picongpu/param/particle.param:
```diff
@@ -0,0 +1,131 @@
+{
+    namespace particles
+    {
+#ifdef PARAM_SINGLE_PARTICLE
```
Your are not using this, aren't you?

steindev commented on share/picongpu/tests/Fusion/include/picongpu/param/particle.param:
Please delete the things definitions you are not using.

steindev commented on share/picongpu/tests/Fusion/include/picongpu/param/png.param:
This is just the standard `png.param` file, isn't it? If so, please remove.

steindev commented on share/picongpu/tests/Fusion/include/picongpu/param/simulation.param:
```diff
@@ -0,0 +1,79 @@
+    {
+        /** Duration of one timestep
+         *  unit: seconds */
+        constexpr float_64 DELTA_T_SI = 1e-15; // 0.1 fs
```
Comment is wrong. This is 1fs. But I think the comment is not required anyways.

steindev commented on share/picongpu/tests/Fusion/include/picongpu/param/simulation.param:
```diff
@@ -0,0 +1,79 @@
+         *
+         * unit: ELEMENTS/m^3
+         */
+        constexpr float_64 BASE_DENSITY_SI = 1e28; // 1e28 particles per m^3
```
The comment is not required.

steindev commented on share/picongpu/tests/Fusion/include/picongpu/param/speciesDefinition.param:
```diff
@@ -0,0 +1,134 @@
+    /*########################### define species #################################*/
+    // doc-include-start: amu-definition
+    // Atomic mass unit in electron masses (amu/me)
+    // This constant allows defining particle masses relative to the electron mass
```
The inverse is already defined in `simulation.unitless`.

For consistency one should use `electronMassAMU` from there.

steindev commented on share/picongpu/tests/Fusion/include/picongpu/param/speciesInitialization.param:
```diff
@@ -0,0 +1,49 @@
+        using InitPipeline = pmacc::mp_list<
+            CreateDensity<densityProfiles::Homogenous, startPosition::Random, PIC_Deuterons>,
+            CreateDensity<densityProfiles::Homogenous, startPosition::Random, PIC_Tritons>,
+
```
2x superfluous white space.

steindev commented on share/picongpu/tests/Fusion/lib/validate.py:
```diff
@@ -0,0 +1,315 @@
+                    print("   PASSED: Within 1% tolerance")
+                else:
+                    print("   FAILED: Outside 1% tolerance")
+                    validation_passed = False
```
I would turn the 'validation_passed' the other way round. Make the standard 'False' and only set to 'True' if all checks are passed. This way you avoid false positives, which can be hard to detect errors (when adding more features later), while false negatives will always result in instant debugging.
